### PR TITLE
remove optimizer_idx arg in manual optimization

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Removed `optimizer_idx` from `training_step` in case of manual optimization since it's redundant in such a case.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -319,8 +319,7 @@ class LitAutoEncoder(pl.LightningModule):
         super().__init__()
         self.automatic_optimization = False
 
-    def training_step(self, batch, batch_idx, optimizer_idx):
-        # access your optimizers with use_pl_optimizer=False. Default is True
```
Comment (awaelchli) [inline README.md:323]: the comment can stay yes?

File: docs/source/starter/new-project.rst [hunk 2]
```diff
@@ -265,8 +265,7 @@ Turn off automatic optimization and you control the train loop!
     def __init__(self):
         self.automatic_optimization = False
 
-    def training_step(self, batch, batch_idx, optimizer_idx):
-        # access your optimizers with use_pl_optimizer=False. Default is True
```
Comment (awaelchli) [inline docs/source/starter/new-project.rst:269]: here too, should be fine to leave the comment?

**Comment by awaelchli** [comment]
@rohitgr7 can we warn that optimizer_idx is None and that it's not appropriate for manual optimization when the user still has it in the signature?

File: README.md [hunk 3]
```diff
@@ -319,8 +319,7 @@ class LitAutoEncoder(pl.LightningModule):
...
+    def training_step(self, batch, batch_idx):
         (opt_a, opt_b) = self.optimizers(use_pl_optimizer=True)
```
Comment (carmocca) [inline README.md:323]: Can we remove these?

```suggestion
        opt_a, opt_b = self.optimizers(use_pl_optimizer=True)
```

File: docs/source/common/lightning_module.rst [hunk 4]
```diff
@@ -952,15 +952,12 @@ When set to ``False``, Lightning does not automate the optimization process. Thi
...
-
 .. code-block:: python
 
     def __init__(self):
         self.automatic_optimization = False
 
-    def training_step(self, batch, batch_idx, optimizer_idx):
-        # access your optimizers with use_pl_optimizer=False. Default is True
+    def training_step(self, batch, batch_idx):
         (opt_a, opt_b) = self.optimizers(use_pl_optimizer=True)
```
Comment (carmocca) [inline docs/source/common/lightning_module.rst:961]:
```suggestion
        opt_a, opt_b = self.optimizers(use_pl_optimizer=True)
```

File: docs/source/common/lightning_module.rst [hunk 5]
```diff
@@ -952,15 +952,12 @@ When set to ``False``, Lightning does not automate the optimization process. Thi
...
 
-In the multi-optimizer case, ignore the ``optimizer_idx`` argument and use the optimizers directly
-
 .. code-block:: python
 
     def __init__(self):
         self.automatic_optimization = False
 
-    def training_step(self, batch, batch_idx, optimizer_idx):
-        # access your optimizers with use_pl_optimizer=False. Default is True
```
Comment (carmocca) [inline docs/source/common/lightning_module.rst:963]: comment

File: tests/core/test_lightning_optimizer.py [hunk 6]
```diff
@@ -89,7 +89,7 @@ def __init__(self):
             super().__init__()
             self.automatic_optimization = False
 
-        def training_step(self, batch, batch_idx, optimizer_idx=None):
+        def training_step(self, batch, batch_idx):
             (opt_1, opt_2) = self.optimizers()
```
Comment (carmocca) [inline tests/core/test_lightning_optimizer.py:93]:
```suggestion
            opt_1, opt_2 = self.optimizers()
```

File: tests/core/test_lightning_optimizer.py [hunk 7]
```diff
@@ -145,7 +145,7 @@ def __init__(self):
...
```
Comment (carmocca) [inline tests/core/test_lightning_optimizer.py:149]:
```suggestion
            opt_1, opt_2 = self.optimizers()
```

File: tests/trainer/optimization/test_manual_optimization.py [hunk 8]
```diff
@@ -40,7 +40,7 @@ def __init__(self):
             super().__init__()
             self.automatic_optimization = False
 
-        def training_step(self, batch, batch_idx, optimizer_idx):
+        def training_step(self, batch, batch_idx):
             # manual
             (opt_a, opt_b) = self.optimizers()
```
Comment (carmocca) [inline tests/trainer/optimization/test_manual_optimization.py:45]: mind removing all these extra parentheses? haha

**Comment by rohitgr7** [comment]
> @rohitgr7 can we warn that optimizer_idx is None and that it's not appropriate for manual optimization when the user still has it in the signature? For backward compatibility right?

@awaelchli Can we deprecate the signature in this case with a warning and remove optimizer_idx? With the old one we can make it None. Wdyt?

**Comment by awaelchli** [comment]
Yes, None or whatever the current value is. I simply suggest to make both

```python
def training_step(self, batch, batch_idx, optimizer_idx)
    ...

```
and 

```python
def training_step(self, batch, batch_idx)
    ...

```
valid signatures (for manual opt) so that switching to 1.3 will not break user code even if they ignored the argument so far.

File: docs/source/common/optimizers.rst [hunk 9]
```diff
@@ -109,6 +117,9 @@ Here is an example on how to use it:
 
     # Scenario for a GAN with gradient accumulation every 2 batches and optimized for multiple gpus.
 
+    def __init__(self):
+        self.automatic_optimization = False
+
     def training_step(self, batch, batch_idx, ...):
```
Comment (tchaton) [inline docs/source/common/optimizers.rst:123]:
```suggestion
    def training_step(self, batch, batch_idx):
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: README.md (hunk 1 above)
Comment (rohitgr7) [reply README.md:323]: comment says to use `use_pl_optimizer=False` but actually it uses `use_pl_optimizer=True` down below.

File: tests/trainer/optimization/test_manual_optimization.py (hunk 8 above)
Comment (rohitgr7) [reply tests/trainer/optimization/test_manual_optimization.py:45]: will do.

File: pytorch_lightning/trainer/training_loop.py [hunk 10]
```diff
@@ -832,14 +832,14 @@ def build_train_args(self, batch, batch_idx, opt_idx, hiddens):
...
             if self.trainer.has_arg("training_step", "optimizer_idx"):
                 args.append(opt_idx)
             else:
                 num_opts = len(self.trainer.optimizers)
                 raise ValueError(
-                    f"Your LightningModule defines {num_opts} optimizers but "
-                    f'training_step is missing the "optimizer_idx" argument.'
+                    f"Your LightningModule defines {num_opts} optimizers but"
+                    f' training_step is missing the "optimizer_idx" argument.'
                 )
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:843]: with how dynamic args are appended here, what's the recommendation for customers who want to strongly type their lightning modules with either mypy or pyre? do we see this style as necessary in order to enable flexibility from the framework side?

Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:843]: will be adding the support for old api with a deprecation warning.

**Review by awaelchli — APPROVED** [review]
nice improvement

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
